### PR TITLE
Add comments about potential race in delta fifo

### PR DESCRIPTION
**Pull request by roycaihw**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

/sig api-machinery

### CODE REVIEW

**Comment by lavalamp** [comment]
@wojtek-t and/or @liggitt can you review this? It would be good not to scare people unnecessarily.

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go [hunk 1]
```diff
@@ -46,6 +46,19 @@ import (
 // TODO(lavalamp): I believe there is a possible race only when using an
 //                 external known object source that the above TODO would
 //                 fix.
+// NOTE: Whether there is a potential race depends on how the comsumer
+//       modifies knownObjects. In Pop(), process function is called under
+//       lock, so it is safe update data structures in it that need to be
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: s/safe update/safe to update/

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go [hunk 2]
```diff
@@ -46,6 +46,19 @@ import (
 // TODO(lavalamp): I believe there is a possible race only when using an
 //                 external known object source that the above TODO would
 //                 fix.
+// NOTE: Whether there is a potential race depends on how the comsumer
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: Can we combine with the TODO, removing the TODO in the process?

Comment (roycaihw) [reply staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: Done.

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go [hunk 3]
```diff
@@ -46,6 +46,20 @@ import (
...
+//
+//                 Example:
+//                 In case of sharedIndexInformer being a consumer
+//                 (https://github.com/kubernetes/kubernetes/blob/0cdd940f/staging/
+//                 src/k8s.io/client-go/tools/cache/shared_informer.go#L192),
+//                 there is no race as knownObjects (s.indexer) is modified safely
+//                 under DeltaFIFO's lock. The only exceptions are GetStore() and
+//                 GetIndexer() methods, which expose ways to modify the underlying
+//                 storage. Currently these two methods are used for creating Lister
+//                 and internal tests.
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: I spend 30 minutes looking into this code and this comment looks correct to me.
However, I would prefer someone else also to take a look.

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go [hunk 4]
```diff
@@ -46,6 +46,20 @@ import (
 // TODO(lavalamp): I believe there is a possible race only when using an
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: Thanks for combining-- maybe we can change this to "NOTE: It is possible to misuse this and cause a race." It's not really a TODO anymore?

Comment (roycaihw) [reply staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: Updated.

**Pull request merged**